core/eth/contracts.go

```diff
@@ -86,9 +101,158 @@ func MustGetV6ContractEventID(name, eventName string) common.Hash {
+		case reflect.Int16:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(int16(num.Int64())))
+
+		case reflect.Int32:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(int32(num.Int64())))
+
+		case reflect.Int64:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(num.Int64()))
+
+		case reflect.Uint8:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(uint8(num.Uint64())))
+
+		case reflect.Uint16:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(uint16(num.Uint64())))
+
+		case reflect.Uint32:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(uint32(num.Uint64())))
+
+		case reflect.Uint64:
+			num := new(big.Int).SetBytes(topics[0][:])
+			field.Set(reflect.ValueOf(num.Uint64()))
+
+		default:
+			// Ran out of plain primitive types, try custom types
+
+			switch field.Type() {
+			case reflectHash: // Also covers all dynamic types
+				field.Set(reflect.ValueOf(topics[0]))
+
+			case reflectAddress:
+				var addr common.Address
+				copy(addr[:], topics[0][common.HashLength-common.AddressLength:])
+				field.Set(reflect.ValueOf(addr))
+
+			case reflectBigInt:
+				num := new(big.Int).SetBytes(topics[0][:])
+				if arg.Type.T == abi.IntTy {
+					if num.Cmp(abi.MaxInt256) > 0 {
+						num.Add(abi.MaxUint256, big.NewInt(0).Neg(num))
+						num.Add(num, big.NewInt(1))
+						num.Neg(num)
+					}
+				}
+				field.Set(reflect.ValueOf(num))
+
+			default:
+				// Ran out of custom types, try the crazies
+				switch {
+				// static byte array
+				case arg.Type.T == abi.FixedBytesTy:
+					reflect.Copy(field, reflect.ValueOf(topics[0][:arg.Type.Size]))
+				default:
+					return fmt.Errorf("unsupported indexed type: %v", arg.Type)
+				}
+			}
+		}
+		topics = topics[1:]
+	}
+	return nil
+}
+
+func capitalise(input string) string {
+	return abi.ToCamelCase(input)
+}
+
+type ConnectedContract interface {
+	Contract
+	Call(result interface{}, methodName string, args ...interface{}) error
+}
+
+type connectedContract struct {
+	Contract
+	ethClient Client
+	address   common.Address
+}
+
+func NewConnectedContract(contract Contract, ethClient Client, address common.Address) ConnectedContract {
+	return &connectedContract{contract, ethClient, address}
+}
+
+func (contract *connectedContract) Call(result interface{}, methodName string, args ...interface{}) error {
+	data, err := contract.EncodeMessageCall(methodName, args...)
+	if err != nil {
+		return errors.Wrap(err, "unable to encode message call")
+	}
+
+	var rawResult hexutil.Bytes
+	callArgs := CallArgs{To: contract.address, Data: data}
+	err = contract.ethClient.Call(&rawResult, "eth_call", callArgs, "latest")
+	if err != nil {
+		return errors.Wrap(err, "unable to call client")
+	}
+
+	err = contract.ABI().Unpack(result, methodName, rawResult)
+	if err != nil {
+		return errors.Wrap(err, "unable to unpack values")
+	}
+	return nil
+}
```

Comment:
small optimization: you can safely do return errors.Wrap(..) here, nil is handled correctly.